File: examples/python/base/service.py
import http.server
import json
from base.trace import RockTrace


class RockApiService:
    @classmethod
    def run(cls, bot, host='127.0.0.1', port=7625):
        _bot = bot
        _trace = RockTrace()

        class RockApiRequestHandler(http.server.BaseHTTPRequestHandler):
            def log_message(self, format, *args):
                return

            def do_GET(self):
                self.send_response(200)
                self.end_headers()
                self.wfile.write(bytes("Hearthrock!", 'utf-8'))
                return

            def do_POST(self):
                content_length = int(self.headers['Content-Length'])
                raw_data = self.rfile.read(content_length)
                json_data = json.loads(raw_data)

                if self.path == "/trace":
                    ret = _trace.trace(json_data)
                elif self.path == "/mulligan":
                    ret = _bot.get_mulligan_action(json_data)
                elif self.path == "/play":
                    ret = _bot.get_play_action(json_data)
                self.send_response(200)
                self.end_headers()
                self.wfile.write(bytes(json.dumps(ret), 'utf-8'))
                return

        httpd = http.server.HTTPServer((host, port), RockApiRequestHandler)

        while True:
            httpd.handle_request()

File: examples/node.js/base/service.js
const trace = require('./trace')
const http = require('http')

/**
 * Api Service implementation.
 */
class RockApiService {
    constructor(bot, host, port) {
        this.bot = bot;
        this.host = host;
        this.port = port;
    }

    run() {
        let server = http.createServer((request, response) => {
            this.doRequest(request, response);
        });

        server.listen(this.port,this.host, (err) => {
            if (err) {
                throw err;
            }

            console.log(`Serving at http://${this.host}:${this.port}`)
        });
    }

    doGet (request, response) {
        response.end('Hearthrock!')
    }

    doPost (request, response) {
        let ret = "";

        let body = "";
        let bot = this.bot;
        request.on('data', function (chunk) {
            body += chunk;
        });
        request.on('end', function () {
            let json = JSON.parse(body);
            if (request.url == "/trace") {
                ret = trace.RockTrace.trace(json);
            } else if (request.url == "/mulligan") {
                ret = bot.get_mulligan_action(json);
            } else if (request.url == "/play") {
                ret = bot.get_play_action(json);
            } else if (request.url == "/report") {
                ret = bot.report(json);
            }

            response.end(JSON.stringify(ret))
        })
    }

    doRequest (request, response) {
        if (request.method == 'GET') {
            this.doGet (request, response);
        } else {
            this.doPost (request, response);
        }
    }

    static run(bot, host, port) {

        let service = new RockApiService(bot, host, port);

        service.run();
    }

}

var exports = module.exports = {
    RockApiService:RockApiService
};